File: Problem Solving/Algorithms/Warmup/A Very Big Sum/aVeryBigSum.scala
import java.io._
import java.util._

object Solution {

    def aVeryBigSum(ar: Array[Long]): Long = {
        //reduceLeft has no initial value
        //return ar.reduceLeft(_+_)
        //foldLeft has initial value
        return ar.foldLeft(0l)(_+_)
    }

    def main(args: Array[String]) {
        val stdin = scala.io.StdIn

        val printWriter = new PrintWriter(sys.env("OUTPUT_PATH"))

        val arCount = stdin.readLine.trim.toInt

        val ar = stdin.readLine.split(" ").map(_.trim.toLong)
        val result = aVeryBigSum(ar)

        printWriter.println(result)

        printWriter.close()
    }
}

File: Core CS/Algorithms/Warmup/A Very Big Sum/aVeryBigSum.kt
import java.io.*
import java.util.*

fun simpleArraySum(ar: Array<Long>): Long {
    //built in sum()
    //return ar.sum()
    
    //stream fold()
    return ar.fold(0L) { acc, e -> acc + e }
}

fun main(args: Array<String>) {
    val scan = Scanner(System.`in`)

    val arCount = scan.nextLine().trim().toLong()

    val ar = scan.nextLine().split(" ").map{ it.trim().toLong() }.toTypedArray()

    val result = simpleArraySum(ar)

    println(result)
}
